Added custom cells to CourseOutline Screen

ehmadzubair opened the pull request:
Four types of Course blocks now have custom cells.
- HTML - `CourseHTMLTableViewCell`
- Video - `CourseVideoTableViewCell`
- Problem - `CourseProblemTableViewCell`
- Unknown - `CourseUnknownTableViewCell`

I was considering SubClassing UITableViewCell to have `setStyles()` , `addSubviews()` & `setConstraints()` methods in the superclass as much as possible. However, the guidelines conflict with that design. Hence, there's a lot of repeated code... Hope it's okay.

The Colours are set from `OEXStyles`. One color is used outside of edX Design Palette (http://ux.edx.org/#colors), The `CourseOutlineFooterView` background color. Should i add the color (neutral-xxlight) to the code?

The init(frame:) method is useless, but it has to be there because we're calling the designated initialiser.

ehmadzubair commented:
@aleffert please review

ehmadzubair commented:
@aleffert please review.
![ios simulator screen shot 15-may-2015 2 28 16 pm](https://cloud.githubusercontent.com/assets/10597112/7650487/6427659a-fb10-11e4-85ac-be342c9ed265.png)

ehmadzubair commented:
@aleffert The first snapshot test that you added yesterday is failing on this one.

aleffert commented:
Yeah. We'll probably need to update it. It's pretty simple. You can just go to the test and do

```
recordMode = true
```

That will generate a new screenshot. Then you can remove that line.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+
+    static let identifier = "CourseHTMLTableViewCellIdentifier"
+    
+    var fontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 15.0)
```
These can all be ``let`` since we don't change these.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        addSubViews()
```
``subviews`` is one word to match UIKit

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+    {
+        fontStyle.applyToLabel(titleLabel)
+        leftImageButton.titleLabel?.font = UIFont.fontAwesomeOfSize(13)
+        leftImageButton.setTitle(String.fontAwesomeIconWithName(.FileO), forState: .Normal)
```
This wasn't in when you first made this PR, so apologies, but can you go through the ``Icon`` class instead? That gives us a chance to abstract out the icon font we use.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+    
+    required init(coder aDecoder: NSCoder) {
```
Let's group this with the other ``init`` near the top

aleffert commented on Source/CourseOutlineFooterView.swift:
```diff
@@ -0,0 +1,115 @@
+    {
+        nextSubSectionView.backgroundColor = OEXStyles.sharedStyles()?.neutralXLight()
+        
+        nextSubSectionLabel.font = UIFont(name: "OpenSans", size: 12.0)
```
These should be using OEXTextStyle

aleffert commented on Source/CourseOutlineFooterView.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    private func addSubviews()
+    {
+        backgroundView = UIView(frame: CGRectMake(0, 0, 320.0, 100.0))
```
We don't want to hardcode the width. Instead it should just pick up the container width.

aleffert commented on Source/CourseOutlineTableSource.swift:
```diff
@@ -53,12 +103,38 @@ class CourseOutlineTableController : UITableViewController {
-            cell.block = nodes[indexPath.row]
-        }
-        return cell
+            println(nodes[indexPath.row].type.displayType.hashValue)
```
This println snuck in.

aleffert commented on Source/CourseOutlineTableSource.swift:
```diff
@@ -53,12 +103,38 @@ class CourseOutlineTableController : UITableViewController {
+            
+
+            }
+    return UITableViewCell();
```
We should probably ``assert`` in this case

aleffert commented on Source/CourseVideoTableViewCell.swift:
```diff
@@ -0,0 +1,111 @@
+    case None
+}
+
+// TODO : Make a property indexPath for the table view cell and then make a delegate which takes the TouchUpInside event as "downloadButtonPressed(indexPath : NSIndexPath)" method in the delegate.
```
Instead of passing the indexPath, maybe just pass the blockID?

aleffert replied:
Or the view itself?

ehmadzubair replied:
I'm going to go with `blockID` on this one.

aleffert commented on Source/OEXStyles.h:
```diff
@@ -46,4 +47,10 @@
 
+#pragma mark Utility
+
+- (nonnull UIColor*)utilityBase;
```
Way to find the right color on the site! Seems like this should be ``utilitySuccessBase``?

ehmadzubair replied:
You're right...

aleffert commented on Source/edX-Info.plist:
```diff
@@ -53,9 +53,11 @@
 	<key>UIAppFonts</key>
 	<array>
+		<string>OpenSans-Bold.ttf</string>
+		<string>OpenSans-Light.ttf</string>
```
Are light and bold definitely in the spec? Last I checked we weren't actually using them.

ehmadzubair replied:
I think the timeLabel looked more accurate when i used OpenSans-Light. That's why i added it. Does the one in the screenshot look ok? It is currently using OpenSans-Regular

aleffert replied:
Let's leave it as regular for now. We'll inevitably do a polish pass with the design team and tweaks will fall out.

aleffert commented on Source/temp.storyboard:
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
??

ehmadzubair replied:
added it for visual testing of autolayout constraints.

aleffert commented on Source/CourseOutlineFooterView.swift:
```diff
@@ -0,0 +1,115 @@
+        
+        subSectionTitleLabel.font = UIFont(name: "OpenSans", size: 13.0)
+        
+        nextButton.setTitle("NEXT", forState: .Normal)
```
This should be normal case ("Next"). I'm not sure it should be capitalized (that's kind of an Android thing), but if it is, we should do the capitalization in code, since all caps is really a style property not a string property. Also, it should be using OEXLocalizedString

aleffert commented:
Just looking at the duplicated code, I think this is definitely a case where it makes sense to use a subclass or have one configurable class that you can set a few properties on.

For example, similar to how UITableViewCell has titleLabel, detailTextLabel, and friends, you could imagine us just having a block cell that has an "icon" property, and a "title" and "subtitle" property. Then each subclass is pretty simple, it just reads the block data and sets the view properties appropriately.

A related solution, is to have a normal UIView subclass with those properties and then each table cell would just add one of those views as a subview and customize it appropriately. Then each cell class can inherit directly from UTableViewCell and basically all the cell classes are responsible for is the model to view binding.

In general, if the main alternative to inheritance is duplicating a lot of code, then inheritance is the way to go. But often there's a design where you can break the pieces apart into a more composable form and in these cases, inheritance is often the easy solution, but not the better solution.

As a broad example, consider the LoadStateViewController. One way to solve the problem of having some standard view loading logic is to have a super class that all your controllers inherit from. But inevitably you end up with a messy complicated super view controller with a lot of coupled functionality and changing it becomes dangerous since every controller inherits from it. Another way to solve this problem, which LoadStateViewController does, is to take the relevant pieces to a particular problem and put them into a class with a simple interface that another controller can instantiate when it needs to. In that case, since there's a small abstraction boundary, it's easy to change.

Ultimately, you need to use your judgment about what makes the best code and if the guidelines suggest otherwise, well, they're just suggestions. Here I don't see a super class getting particularly complicated or nasty so it's probably fine.

aleffert commented on Source/CourseOutlineFooterView.swift:
```diff
@@ -0,0 +1,115 @@
+        previousButton.setTitleColor(OEXStyles.sharedStyles()?.primaryBaseColor(), forState: .Normal)
+        previousButton.titleLabel?.font = UIFont(name: "OpenSans-Semibold", size: 13.0)
+        
+        nextSubSectionLabel.text = "Next Subsection"
```
I'm not sure we're ready to support the next section stuff yet. Maybe we can save it on a branch or something so it's there when we need it.

aleffert commented on Source/CourseOutlineTableSource.swift:
```diff
@@ -45,6 +50,51 @@ class CourseOutlineTableController : UITableViewController {
+        }
+    }
+    
+    override func tableView(tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
```
Let's not worry about this for now.

aleffert commented on Source/CourseOutlineTableSource.swift:
```diff
@@ -45,6 +50,51 @@ class CourseOutlineTableController : UITableViewController {
+            switch childNodes[indexPath.row].type{
+            case .HTML:
+                return 60.0
+            case .Video:
```
If they're all the same, you can use ``fallthrough``. But probably we should have the height be a property of the cell class, so it's closer to the view.

It might also be worth leaving a comment about switching to automatic cell heights when we drop iOS 7 support.

ehmadzubair replied:
How do i go about putting it as a property of `UITableViewCell` subclass?

aleffert replied:
yep!

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+        leftImageButton.snp_makeConstraints { (make) -> Void in
+            make.leading.equalTo(self).offset(20)
+            make.centerY.equalTo(self)
+            make.size.equalTo(CGSizeMake(25, 25))
```
We should name these constants so it's easier to read. And if there are ones we probably want to use elsewhere we should figure out a good place to put them.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+    static let identifier = "CourseHTMLTableViewCellIdentifier"
+    
+    var fontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 15.0)
+    var leftImageButton = UIButton()
```
We should pass a button type to the UI button initializer. If you use the default type, then it handles fading the button on touchdown for you.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    var fontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 15.0)
+    var leftImageButton = UIButton()
+    var titleLabel = UILabel()
```
I'm not sure it matters for these classes, but we should probably go through init(frame:) since it's the designated initializer.

ehmadzubair replied:
Should i do it in this PR?

aleffert replied:
Let's only worry about it for UIButton, since I know that has different behavior

aleffert commented:
Main things to do are: reduce the duplicated code, make sure to use text styles, and name some constants.

ehmadzubair commented:
@aleffert please review

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+
+import UIKit
+
+class CourseContentView: UIView {
```
Let's call this ``CourseOutlineItemView`` that way it's clear that it's scoped to the course outline.

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+        super.init(frame: CGRectMake(0, 0, 0, 0))
+        
+        fontStyle.applyToLabel(titleLabel)
+        titleLabel.text = title ?? ""
```
Since ``title`` is not nullable, you don't need the ``??`` part.

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+class CourseContentView: UIView {
+
+    let fontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 15.0)
+    let smallerFontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 13.0)
```
Let's call this ``detailFontStyle``

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+        
+        smallerFontStyle.applyToLabel(subtitleLabel)
+        subtitleLabel.text = subtitle ?? ""
+        subtitleLabel.textColor = OEXStyles.sharedStyles()?.neutralBase()
```
Color should be part of the text style, this way it's all grouped in one place.

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+    let fontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 15.0)
+    let smallerFontStyle = OEXTextStyle(font: OEXTextFont.ThemeSans, size: 13.0)
+    var titleLabel = UILabel()
+    var leftImageButton = UIButton.buttonWithType(UIButtonType.Custom) as! UIButton
```
Instead of ``left`` and ``right``, these should be ``leading`` and ``trailing`` or maybe ``primary`` and ``secondary``. This way it makes sense with RTL.

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+    var subtitleLabel = UILabel()
+    
+    init(title : String, subtitle : String, leftImageIcon : Icon, rightImageIcon : Icon?){
+        super.init(frame: CGRectMake(0, 0, 0, 0))
```
Might be slightly cleaner to use ``CGRectZero`` instead of ``CGRectMake(0, 0, 0, 0)``

aleffert commented on Source/CourseContentView.swift:
```diff
@@ -0,0 +1,101 @@
+    
+    func setLeftIconColor(color : UIColor!)
+    {
+        leftImageButton.setTitleColor(color, forState: .Normal)
```
Might be a little nicer to make this a computed property:
```
var leftIconColor : UIColor {
  get {
    return leftImageButton.titleColorForState(.Normal)
  }
  set {
    leftImageButton.setTitleColor(color, forState:.Normal)
  }
}
```

aleffert replied:
Do the pressed states look right?

ehmadzubair replied:
Yes it does...

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,42 @@
+    
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        addSubview(content)
```
You should add children to the cell's contentView instead of the cell directly.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,42 @@
+        addSubview(content)
+        content.snp_makeConstraints { (make) -> Void in
+            make.top.equalTo(self)
+            make.bottom.equalTo(self)
```
You can just do:
```
make.edges.equalTo(self)
```

aleffert commented on Source/CourseOutlineHeaderCell.swift:
```diff
@@ -12,9 +12,61 @@ import UIKit
+        backgroundView = UIView(frame: CGRectMake(0, 0, 320.0, 48.0))
+        backgroundView?.backgroundColor = UIColor.whiteColor()
+        
+        headerLabel.font = UIFont(name: "OpenSans", size: 13.0)
```
Should be using OEXTextStyle

aleffert commented on Source/CourseOutlineHeaderCell.swift:
```diff
@@ -12,9 +12,61 @@ import UIKit
+    }
+    
+    private func setStyles(){
+        backgroundView = UIView(frame: CGRectMake(0, 0, 320.0, 48.0))
```
The header cell automatically resizes its backgroundView's width, right? Might be worth leaving a comment about that. The hardcoded 320 makes it look like it's making assumptions about screen width that aren't valid.

ehmadzubair replied:
Yes it does and okay... Will do it... This definitely makes more sense

aleffert commented on Source/CourseOutlineHeaderCell.swift:
```diff
@@ -12,9 +12,61 @@ import UIKit
+    
+    private func setStyles(){
+        backgroundView = UIView(frame: CGRectMake(0, 0, 320.0, 48.0))
+        backgroundView?.backgroundColor = UIColor.whiteColor()
```
I just added OEXStyles.standardBackgroundColor

aleffert commented on Source/CourseUnknownTableViewCell.swift:
```diff
@@ -0,0 +1,40 @@
+
+    static let identifier = "CourseUnknownTableViewCellIdentifier"
+    
+    var content = CourseContentView(title: "", subtitle: "Test Subtitle Text", leftImageIcon: Icon.CourseUnknownContent, rightImageIcon: nil)
```
Worth adding a ``TODO`` here for the item text.

aleffert commented on Source/OEXConfig.m:
```diff
@@ -173,7 +173,7 @@ - (OEXZeroRatingConfig*)zeroRatingConfig {
 
 - (BOOL)shouldEnableNewCourseNavigation {
-    return NO;
+    return YES;
```
This shouldn't go in

aleffert commented:
There are a bunch of small things left, but this looks good! Also, it doesn't matter too much, but there are a bunch of ``var``s that be ``let``s and a lot of things that should probably be marked as private.

ehmadzubair commented:
@aleffert Did the following.

-Replaced unnecessary vars with lets and added relevant comments
- Set relevant functions private.
- Removed unnecessary fonts.
- Fixed the config

Will rebase the withoutFooterView branch with this branch and merge it into master whenever it's ready.

aleffert commented on Source/CourseHTMLTableViewCell.swift:
```diff
@@ -0,0 +1,39 @@
+        contentView.addSubview(content)
+        content.snp_makeConstraints { (make) -> Void in
+            make.edges.equalTo(contentView)
+            make.center.equalTo(contentView)
```
There is only one possibility for center if the edges are set, so you don't need to set it here.

aleffert commented on Source/CourseOutlineItemView.swift:
```diff
@@ -0,0 +1,111 @@
+    
+    let subtitleLabel = UILabel()
+    
+    init(title : String, subtitle : String, leftImageIcon : Icon, rightImageIcon : Icon?){
```
this also needs to be leading and trailing instead of left and right.

aleffert commented on Source/CourseOutlineFooterView.swift:
```diff
@@ -0,0 +1,118 @@
+        
+        nextButton.setTitle("NEXT", forState: .Normal)
+        nextButton.setTitleColor(OEXStyles.sharedStyles()?.primaryBaseColor(), forState: .Normal)
+        nextButton.titleLabel?.font = UIFont(name: "OpenSans-Semibold", size: 13.0)
```
This needs to be using OEXTextStyle

aleffert replied:
Oh nevermind, you're taking this part out for now, right?

ehmadzubair replied:
Yes, i am

aleffert commented on Source/CourseOutlineHeaderCell.swift:
```diff
@@ -12,9 +12,63 @@ import UIKit
+            make.top.equalTo(self.snp_top)
+            make.leading.equalTo(self.snp_leading)
+            make.trailing.equalTo(self.snp_trailing)
+            make.height.equalTo(0.5)
```
there's now OEXStyles.dividerHeight

aleffert commented on UIConstants.swift:
```diff
@@ -0,0 +1,20 @@
+class UIConstants: NSObject {
+
+    //MARK: - AutoLayout Constants
+    static let ALIconSize = CGSizeMake(25, 25)
```
I'm not sure the "AL" is useful. If we're using autolayout or not, the constants are the same.

Also, it seems like these are specific to the CourseOutline cells. Though, we'll need to standardize on margins more in the future.

Anyway, since these are all local to CourseOutlineCellView let's just put them in that file.

aleffert commented:
A few very small changes left. Make those and you're good to merge :+1:
